src/psyclone/docstring_parser.py

```diff
@@ -0,0 +1,539 @@
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors: A. B. G. Chalk, STFC Daresbury Lab
+# -----------------------------------------------------------------------------
+#
+# This code is influenced by the docstring_parser packaged created by Marcin
+# Kurczewski. Their licence is:
+#
+# The MIT License (MIT)
+# Copyright (c) 2018 Marcin Kurczewski
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to
+# deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense,
+# and/or sell copies of the Software, and to permit persons to whom the
+# Software is furnished to do so, subject to the following conditions:
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+'''This module contains the PSyclone docstring parsing code. This
+implementation influenced by the docstring_parser package created by
+Marcin Kurczewski - https://github.com/rr-/docstring_parser.
+'''
+
+from collections import OrderedDict
+from dataclasses import dataclass
+import inspect
+import os
+import re
+from typing import Any, List, Union
+
+from psyclone.errors import DocParseError
+
+try:
+    from sphinx.util.typing import stringify_annotation
+except ImportError:
+    # Fix for Python-3.7 where sphinx didn't yet rename this.
+    # TODO 2837: Can remove this 3.7 sphinx import
+    try:
+        from sphinx.util.typing import stringify as stringify_annotation
+    # Igoring coverage from the no sphinx workaround as too difficult to do
+    except ImportError:  # pragma: no cover
+        from psyclone.utils import stringify_annotation  # pragma: no cover
+
+
+@dataclass
+class ArgumentData():
+    ''' Dataclass object to represent  a :param...: docstring element.
+
+    :param str name: The name of the parameter.
+    :param str datatype: The string representation of the parameter's
+                         datatype.
+    :param str desc: The description of the parameter.
+    :param bool inline_type: whether the type of this parameter should be
+                             inline or a separate :type param: docstring line.
+    '''
+    name: str
+    datatype: str
+    desc: str
+    inline_type: bool
```

Comment:
Since this class represents the description of an argument, could it also store the argument object to save you having to pass it to `gen_docstring()`?

Reply:
I think no - the `ArgumentData` can be computed from `superclass.apply` and later applied to `class.apply` when inheriting docstrings.